gcp/gcp.go

```diff
@@ -487,6 +487,93 @@ const (
 	Owner   Role = "OWNER"
 )
 
+// Search calls google.com/cloudprint/search to query printers owned by the user/robot.
+func (gcp *GoogleCloudPrint) Search(q, atype, connection_status string) (map[string]lib.Printer, error) {
+	var client *http.Client
+	if gcp.userClient != nil {
+		client = gcp.userClient
+	} else {
+		client = gcp.robotClient
+	}
+	form := url.Values{}
+	if q != "" {
+		form.Set("q", q)
+	}
+	if atype != "" {
+		form.Set("type", atype)
+	}
+	if connection_status != "" {
+		form.Set("connection_status", connection_status)
+	}
+	form.Set("extra_fields", "semanticState")
+
+	responseBody, _, _, err := postWithRetry(client, gcp.baseURL+"search", form)
+	if err != nil {
+		return nil, err
+	}
+
+	printers := make(map[string]lib.Printer)
+
+	var printersData struct {
+		Printers []struct {
```

Comment:
You should factor this definition out of this and the `Printer` method and unify them. It can be a private type within this file.